File: Javascript(一个前端工程师的挣扎）/Classic Questions/139.word_break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */

// DFS解法 -- timeout
var wordBreak = function(s, wordDict) {
    if (s === null || s === '') return true;
    let dic = new Set(wordDict);

    return dfsHelper(dic, s, 0);
};

const dfsHelper = function (dic, s, curIdx) {
    if (curIdx === s.length) return true;

    let res = false;
    for (let i = curIdx; i <= s.length; i++) {
        let word = s.substring(curIdx, i)

        if (dic.has(word)) {
            res = res || dfsHelper(dic, s, curIdx + i);
        }
    }

    return res;
}

// DFS with memorization
var wordBreak = function(s, wordDict) {
    if (s === null || s === '') return true;
    let dic = new Set(wordDict);
    let mem = new Array(s.length).fill(null);
    return dfsHelper(dic, s, 0, mem);
};

const dfsHelper = function (dic, s, curIdx, mem) {
    if (curIdx === s.length) return true;

    if (mem[curIdx] != null) {
        return mem[curIdx];
    }

    for (let i = curIdx; i <= s.length; i++) {
        let word = s.substring(curIdx, i)

        if (dic.has(word) && dfsHelper(dic, s, i, mem)) {
           return mem[curIdx] = true;
        }
    }

    return mem[curIdx] = false;
}

// BFS with memorization
var wordBreak = function(s, wordDict) {
    if (s === null || s.length === 0) return false;
    let dic = new HashSet(wordDict);
    let queue = [];
    let visited = new Array(s.length).fill(0);

    queue.push(0);
    while (queue.length > 0) {
        let curIdx = queue.shift();

        if (visited[curIdx] === 1) continue;

        for (let i = curIdx + 1; i < s.length; i++) {
            if (dic.has(s.substring(curIdx, i))) {
                queue.push(i);

                if (i === s.length) return true;
            }
        }

        visited[curIdx] = 1;
    }

    return false;

};

// Dynamic Programming
// state function: if dp[i] = true && wordDic.has(i, j) => dp[j] = true
var wordBreak = function (s, wordDict) {
    if (s === null || s.length === 0) return false;

    let dic = new HashSet(wordDict);
    let dp = new Array(s.length + 1);
    dp[0] = true;

    for (let i = 1; i <= s.length; i++) {
        for (let j = 0; j < i; j++) {
            if (dp[j] && dic.has(s.substring(j, i))) dp[i] = true;
            break;
        }
    }

    return dp[s.length];

}

File: Java/139.word_break.java
// Intuitive DFS Solution
// 做一个注释，这个简单的解法写的有点磕绊，解法本身会超时
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        HashSet<String> dic = new HashSet<>();
        HashSet<String> valid = new HashSet<>();
        for (String word : wordDict) {
            dic.add(word);
        }

        return dfsHelper(s, dic, 0, visited, "");
    }
    public boolean dfsHelper(String s, HashSet<String> dic, int index, HashSet<String> visited, String cur) {
        if (index == s.length()) return true;
        String remainingString = s.substring(index, s.length());
        if (visited.contains(remainingString)) return true;

        boolean res = false;
        for (int i = index; i <= s.length(); i++) {
            String tempWord = s.substring(index, i);
            if (dic.contains(tempWord)) {
                res = res || dfsHelper(s, dic, i, visited, cur + tempWord);
            }
        }
        return res;
    }
}

// BFS Solution
// 不知道为啥这个解法写了半天才解出来
// BFS剪枝的方法还是要记得的

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(0);
        int[] visited = new int[s.length()];

        HashSet<String> dic = new HashSet<>();
        for (String word : wordDict) {
            dic.add(word);
        }

        while (!queue.isEmpty()) {
            int index = queue.poll();

            if (visited[index] == 0) {
                for (int end = index + 1; end <= s.length(); end++) {
                    if (dic.contains(s.substring(index, end))) {
                        if (end == s.length()) return true;
                        queue.add(end);
                    }
                }
            }
            visited[index] = 1;
        }
        return false;
    }
}
